pallets/contracts/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+
+        // Simply forwards to the `update_schedule` function in the Contract module.
+        #[weight = pallet_contracts::Call::<T>::update_schedule(schedule).get_dispatch_info().weight]
+        pub fn update_schedule(origin, schedule: Schedule) -> DispatchResult {
+            <pallet_contracts::Module<T>>::update_schedule(origin, schedule)
+        }
+
+        /// Simply forwards to the `put_code` function in the Contract module.
+        ///
+        /// # Additional functionality
+        /// 1. Allow origin to pass some meta-details related to template code.
+        /// 2. Charge protocol fee for deploying the template.
+        #[weight = 50_000_000.saturating_add(pallet_contracts::Call::<T>::put_code(code.clone()).get_dispatch_info().weight)]
+        pub fn put_code(
+            origin,
+            meta_info: SmartExtensionMetadata<BalanceOf<T>>,
+            code: Vec<u8>
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+
+            // Save metadata related to the SE template
+            // Generate the code_hash here as well because there is no way
+            // to read it directly from the upstream `pallet-contracts` module.
+            let code_hash = T::Hashing::hash(&code);
+
+            // Call underlying function
+            <pallet_contracts::Module<T>>::put_code(origin, code)?;
+
+            // Charge the protocol fee
+            T::ProtocolFee::charge_fee(ProtocolOp::ContractsPutCode)?;
+            <TemplateMetaDetails<T>>::insert(code_hash, TemplateMetadata {
+                meta_info: meta_info,
+                owner: sender,
+                frozen: false
+            });
+            Ok(())
+        }
+
+        // Simply forwards to the `call` function in the Contract module.
+        #[weight = pallet_contracts::Call::<T>::call(dest, value, gas_limit, data.clone()).get_dispatch_info().weight]
+        pub fn call(
+            origin,
+            dest: <T::Lookup as StaticLookup>::Source,
+            #[compact] value: BalanceOf<T>,
+            #[compact] gas_limit: Gas,
+            data: Vec<u8>
+        ) -> DispatchResultWithPostInfo {
+            <pallet_contracts::Module<T>>::call(origin, dest, value, gas_limit, data)
+        }
+
+        /// Simply forwards to the `instantiate` function in the Contract module.
+        ///
+        /// # Additional functionality
+        /// 1. Check whether instantiation of given code_hash is allowed or not.
+        /// 2. Charge instantiation fee.
+        ///
+        /// # Errors
+        /// InstantiationIsNotAllowed - It occurred when instantiation of the template is frozen.
+        #[weight = 500_000_000 + *gas_limit]
+        pub fn instantiate(
+            origin,
+            #[compact] endowment: BalanceOf<T>,
+            #[compact] gas_limit: Gas,
+            code_hash: CodeHash<T>,
+            data: Vec<u8>
+        ) -> DispatchResultWithPostInfo {
```

Comment:
Should the user specify the (maximum) instantiation fee they are willing to pay (to avoid any front-running type problems where the fee is changed just before the contract is instantiated)?

Reply:
I don't think the front-run will be an issue here as we don't support tipping. This issue may come up when the transaction submitted by the user to a networked delayed node or due to some reasons transaction priority will go down. 

In general, front-running will not occur but to be defensive I can add the maximum instantiation_fee if you want.